File: rust/crates/storage/src/connection_repo.rs
use std::sync::Arc;

use chrono::{DateTime, Utc};
use parking_lot::Mutex;
use polydb_core::{
    ConnectionId, ConnectionInfo, CoreError, CoreResult, CreateConnectionRequest, DatabaseKind,
    UpdateConnectionRequest,
};
use rusqlite::Connection as RusqliteConnection;

pub struct ConnectionRepository {
    conn: Arc<Mutex<RusqliteConnection>>,
}

impl ConnectionRepository {
    pub(crate) fn new(conn: Arc<Mutex<RusqliteConnection>>) -> Self {
        Self { conn }
    }

    pub fn create(&self, req: &CreateConnectionRequest) -> CoreResult<ConnectionInfo> {
        let conn = self.conn.lock();
        let id = ConnectionId::new_v4();
        let now = Utc::now();
        let options_json = serde_json::to_string(&req.options)
            .map_err(|e| CoreError::Storage(format!("serialize options: {e}")))?;
        let ssh_json = req
            .ssh_tunnel
            .as_ref()
            .map(serde_json::to_string)
            .transpose()
            .map_err(|e| CoreError::Storage(format!("serialize ssh: {e}")))?;
        let kind_str = req.kind.to_string();

        conn.execute(
            "INSERT INTO connections (id, name, kind, host, port, database, username, password_ref, options, ssh_tunnel, default_schema, read_only, created_at, updated_at)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14)",
            rusqlite::params![
                id.to_string(),
                req.name,
                kind_str,
                req.host,
                req.port.map(|p| p as i64),
                req.database,
                req.username,
                req.password_ref,
                options_json,
                ssh_json,
                req.default_schema,
                req.read_only.unwrap_or(false),
                now.to_rfc3339(),
                now.to_rfc3339(),
            ],
        ).map_err(|e| CoreError::Storage(format!("insert connection: {e}")))?;

        Ok(ConnectionInfo {
            id,
            name: req.name.clone(),
            kind: req.kind,
            host: req.host.clone(),
            port: req.port,
            database: req.database.clone(),
            username: req.username.clone(),
            options: req.options.clone(),
            ssh_tunnel: req.ssh_tunnel.clone(),
            default_schema: req.default_schema.clone(),
            read_only: req.read_only,
            created_at: now,
            updated_at: now,
        })
    }

    pub fn list(&self) -> CoreResult<Vec<ConnectionInfo>> {
        let conn = self.conn.lock();
        let mut stmt = conn.prepare(
            "SELECT id, name, kind, host, port, database, username, options, ssh_tunnel, default_schema, read_only, created_at, updated_at FROM connections ORDER BY created_at"
        ).map_err(|e| CoreError::Storage(format!("list connections prepare: {e}")))?;

        let rows = stmt
            .query_map([], |row| {
                let id_str: String = row.get(0)?;
                let name: String = row.get(1)?;
                let kind_str: String = row.get(2)?;
                let host: Option<String> = row.get(3)?;
                let port: Option<i64> = row.get(4)?;
                let database: Option<String> = row.get(5)?;
                let username: Option<String> = row.get(6)?;
                let options_json: String = row.get(7)?;
                let ssh_json: Option<String> = row.get(8)?;
                let default_schema: Option<String> = row.get(9)?;
                let read_only: i64 = row.get(10)?;
                let created_at_str: String = row.get(11)?;
                let updated_at_str: String = row.get(12)?;
                Ok((
                    id_str,
                    name,
                    kind_str,
                    host,
                    port,
                    database,
                    username,
                    options_json,
                    ssh_json,
                    default_schema,
                    read_only,
                    created_at_str,
                    updated_at_str,
                ))
            })
            .map_err(|e| CoreError::Storage(format!("list connections query: {e}")))?;

        let mut result = Vec::new();
        for row in rows {
            let (
                id_str,
                name,
                kind_str,
                host,
                port,
                database,
                username,
                options_json,
                ssh_json,
                default_schema,
                read_only,
                created_at_str,
                updated_at_str,
            ) = row.map_err(|e| CoreError::Storage(format!("row: {e}")))?;

            let id: ConnectionId = id_str
                .parse()
                .map_err(|e| CoreError::Storage(format!("parse id: {e}")))?;
            let kind = parse_db_kind(&kind_str)?;
            let options: std::collections::HashMap<String, String> =
                serde_json::from_str(&options_json).unwrap_or_default();
            let ssh_tunnel = ssh_json
                .as_deref()
                .map(serde_json::from_str)
                .transpose()
                .map_err(|e| CoreError::Storage(format!("parse ssh: {e}")))?;
            let created_at: DateTime<Utc> = created_at_str
                .parse()
                .map_err(|e| CoreError::Storage(format!("parse created_at: {e}")))?;
            let updated_at: DateTime<Utc> = updated_at_str
                .parse()
                .map_err(|e| CoreError::Storage(format!("parse updated_at: {e}")))?;

            result.push(ConnectionInfo {
                id,
                name,
                kind,
                host,
                port: port.map(|p| p as u16),
                database,
                username,
                options,
                ssh_tunnel,
                default_schema,
                read_only: bool_from_stored(read_only),
                created_at,
                updated_at,
            });
        }
        Ok(result)
    }

    pub fn get(&self, id: ConnectionId) -> CoreResult<Option<ConnectionInfo>> {
        let conn = self.conn.lock();
        let mut stmt = conn.prepare(
            "SELECT id, name, kind, host, port, database, username, options, ssh_tunnel, default_schema, read_only, created_at, updated_at FROM connections WHERE id = ?1"
        ).map_err(|e| CoreError::Storage(format!("get connection prepare: {e}")))?;

        let result = stmt.query_row(rusqlite::params![id.to_string()], |row| {
            let id_str: String = row.get(0)?;
            let name: String = row.get(1)?;
            let kind_str: String = row.get(2)?;
            let host: Option<String> = row.get(3)?;
            let port: Option<i64> = row.get(4)?;
            let database: Option<String> = row.get(5)?;
            let username: Option<String> = row.get(6)?;
            let options_json: String = row.get(7)?;
            let ssh_json: Option<String> = row.get(8)?;
            let default_schema: Option<String> = row.get(9)?;
            let read_only: i64 = row.get(10)?;
            let created_at_str: String = row.get(11)?;
            let updated_at_str: String = row.get(12)?;
            Ok((
                id_str,
                name,
                kind_str,
                host,
                port,
                database,
                username,
                options_json,
                ssh_json,
                default_schema,
                read_only,
                created_at_str,
                updated_at_str,
            ))
        });

        match result {
            Ok((
                id_str,
                name,
                kind_str,
                host,
                port,
                database,
                username,
                options_json,
                ssh_json,
                default_schema,
                read_only,
                created_at_str,
                updated_at_str,
            )) => {
                let id: ConnectionId = id_str
                    .parse()
                    .map_err(|e| CoreError::Storage(format!("parse id: {e}")))?;
                let kind = parse_db_kind(&kind_str)?;
                let options: std::collections::HashMap<String, String> =
                    serde_json::from_str(&options_json).unwrap_or_default();
                let ssh_tunnel = ssh_json
                    .as_deref()
                    .map(serde_json::from_str)
                    .transpose()
                    .map_err(|e| CoreError::Storage(format!("parse ssh: {e}")))?;
                let created_at: DateTime<Utc> = created_at_str
                    .parse()
                    .map_err(|e| CoreError::Storage(format!("parse created_at: {e}")))?;
                let updated_at: DateTime<Utc> = updated_at_str
                    .parse()
                    .map_err(|e| CoreError::Storage(format!("parse updated_at: {e}")))?;

                Ok(Some(ConnectionInfo {
                    id,
                    name,
                    kind,
                    host,
                    port: port.map(|p| p as u16),
                    database,
                    username,
                    options,
                    ssh_tunnel,
                    default_schema,
                    read_only: bool_from_stored(read_only),
                    created_at,
                    updated_at,
                }))
            }
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
            Err(e) => Err(CoreError::Storage(format!("get connection: {e}"))),
        }
    }

    pub fn update(
        &self,
        id: ConnectionId,
        req: &UpdateConnectionRequest,
    ) -> CoreResult<Option<ConnectionInfo>> {
        let existing = self.get(id)?;
        let Some(existing) = existing else {
            return Ok(None);
        };

        let now = Utc::now();

        // password_ref 不随 ConnectionInfo 下发：更新请求未带新 ref 时保留库中已有 ref
        // （与 Go 侧 repo.Update 对齐，避免改名/改 host 等更新把密码引用清空）。
        // 必须在获取锁之前查询：get_password_ref 需要同一把锁，而 parking_lot 不可重入，
        // 持锁时调用会自死锁。
        let password_ref = match req.password_ref.as_ref() {
            Some(r) => Some(r.clone()),
            None => self.get_password_ref(id)?,
        };

        let conn = self.conn.lock();

        let name = req.name.as_deref().unwrap_or(&existing.name);
        let host = req.host.as_ref().or(existing.host.as_ref());
        let port = req.port.or(existing.port);
        let database = req.database.as_ref().or(existing.database.as_ref());
        let username = req.username.as_ref().or(existing.username.as_ref());
        let options = req.options.as_ref().unwrap_or(&existing.options);
        let ssh_tunnel = req.ssh_tunnel.as_ref().or(existing.ssh_tunnel.as_ref());
        let default_schema = req
            .default_schema
            .as_ref()
            .or(existing.default_schema.as_ref());
        let read_only = req.read_only.or(existing.read_only);

        let options_json = serde_json::to_string(options)
            .map_err(|e| CoreError::Storage(format!("serialize options: {e}")))?;
        let ssh_json = ssh_tunnel
            .map(serde_json::to_string)
            .transpose()
            .map_err(|e| CoreError::Storage(format!("serialize ssh: {e}")))?;

        conn.execute(
            "UPDATE connections SET name=?1, host=?2, port=?3, database=?4, username=?5, password_ref=?6, options=?7, ssh_tunnel=?8, default_schema=?9, read_only=?10, updated_at=?11 WHERE id=?12",
            rusqlite::params![
                name,
                host,
                port.map(|p| p as i64),
                database,
                username,
                password_ref.as_deref(),
                options_json,
                ssh_json,
                default_schema,
                read_only.unwrap_or(false),
                now.to_rfc3339(),
                id.to_string(),
            ],
        ).map_err(|e| CoreError::Storage(format!("update connection: {e}")))?;

        Ok(Some(ConnectionInfo {
            id,
            name: name.to_string(),
            kind: existing.kind,
            host: host.cloned(),
            port,
            database: database.cloned(),
            username: username.cloned(),
            options: options.clone(),
            ssh_tunnel: ssh_tunnel.cloned(),
            default_schema: default_schema.cloned(),
            read_only,
            created_at: existing.created_at,
            updated_at: now,
        }))
    }

    pub fn delete(&self, id: ConnectionId) -> CoreResult<bool> {
        let conn = self.conn.lock();
        let affected = conn
            .execute(
                "DELETE FROM connections WHERE id=?1",
                rusqlite::params![id.to_string()],
            )
            .map_err(|e| CoreError::Storage(format!("delete connection: {e}")))?;
        Ok(affected > 0)
    }

    /// 返回连接存储的 password_ref。该引用不随 ConnectionInfo 下发（红线），
    /// 仅 Connect 时单独查询，再到 keyring 取明文。
    pub fn get_password_ref(&self, id: ConnectionId) -> CoreResult<Option<String>> {
        let conn = self.conn.lock();
        match conn.query_row(
            "SELECT password_ref FROM connections WHERE id = ?1",
            rusqlite::params![id.to_string()],
            |row| row.get::<_, Option<String>>(0),
        ) {
            Ok(v) => Ok(v),
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
            Err(e) => Err(CoreError::Storage(format!("get password_ref: {e}"))),
        }
    }
}

fn parse_db_kind(s: &str) -> CoreResult<DatabaseKind> {
    match s {
        "sqlite" => Ok(DatabaseKind::Sqlite),
        "mysql" => Ok(DatabaseKind::MySql),
        "postgres" => Ok(DatabaseKind::Postgres),
        "mssql" => Ok(DatabaseKind::Mssql),
        "oracle" => Ok(DatabaseKind::Oracle),
        "redis" => Ok(DatabaseKind::Redis),
        _ => Err(CoreError::Storage(format!("unknown database kind: {s}"))),
    }
}

/// read_only 列以 0/1 存储；0 表示"未设置"（线上省略该字段），1 表示 true。
fn bool_from_stored(v: i64) -> Option<bool> {
    if v != 0 {
        Some(true)
    } else {
        None
    }
}

File: go/pkg/storage/connection_repo.go
package storage

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/polydb/polydb/pkg/protocol"
)

// ConnectionRepository 提供 connections 表的 CRUD，密码只落 password_ref 不落明文。
type ConnectionRepository struct {
	db *sql.DB
}

const connColumns = "id, name, kind, host, port, database, username, options, ssh_tunnel, default_schema, read_only, \"group\", color, created_at, updated_at"

func NewConnectionRepository(db *sql.DB) *ConnectionRepository {
	return &ConnectionRepository{db: db}
}

func (r *ConnectionRepository) Create(req *protocol.CreateConnectionRequest) (protocol.ConnectionInfo, error) {
	now := time.Now().UTC()
	id := uuid.NewString()
	optionsJSON, err := json.Marshal(req.Options)
	if err != nil {
		return protocol.ConnectionInfo{}, fmt.Errorf("marshal options: %w", err)
	}
	var sshJSON []byte
	if req.SSHTunnel != nil {
		sshJSON, err = json.Marshal(req.SSHTunnel)
		if err != nil {
			return protocol.ConnectionInfo{}, fmt.Errorf("marshal ssh: %w", err)
		}
	}
	_, err = r.db.Exec(
		`INSERT INTO connections (id, name, kind, host, port, database, username, password_ref, options, ssh_tunnel, default_schema, read_only, "group", color, created_at, updated_at)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		id, req.Name, string(req.Kind), nullStr(req.Host), nullInt(req.Port),
		nullStr(req.Database), nullStr(req.Username), nullStr(req.PasswordRef),
		string(optionsJSON), nullBytes(sshJSON), nullStr(req.DefaultSchema), boolInt(req.ReadOnly),
		nullStr(derefStr(req.Group)), nullStr(derefStr(req.Color)),
		now.Format(time.RFC3339), now.Format(time.RFC3339),
	)
	if err != nil {
		return protocol.ConnectionInfo{}, fmt.Errorf("insert connection: %w", err)
	}
	return protocol.ConnectionInfo{
		ID:            id,
		Name:          req.Name,
		Kind:          req.Kind,
		Host:          req.Host,
		Port:          req.Port,
		Database:      req.Database,
		Username:      req.Username,
		Options:       req.Options,
		SSHTunnel:     req.SSHTunnel,
		DefaultSchema: req.DefaultSchema,
		ReadOnly:      req.ReadOnly,
		Group:         req.Group,
		Color:         req.Color,
		CreatedAt:     now,
		UpdatedAt:     now,
	}, nil
}

func (r *ConnectionRepository) List() ([]protocol.ConnectionInfo, error) {
	rows, err := r.db.Query("SELECT " + connColumns + " FROM connections ORDER BY created_at")
	if err != nil {
		return nil, fmt.Errorf("list connections: %w", err)
	}
	defer rows.Close()
	out := make([]protocol.ConnectionInfo, 0)
	for rows.Next() {
		info, err := scanConnection(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, info)
	}
	return out, rows.Err()
}

func (r *ConnectionRepository) Get(id string) (protocol.ConnectionInfo, error) {
	row := r.db.QueryRow("SELECT "+connColumns+" FROM connections WHERE id = ?", id)
	info, err := scanConnection(row)
	if err == sql.ErrNoRows {
		return protocol.ConnectionInfo{}, ErrNotFound
	}
	return info, err
}

func (r *ConnectionRepository) Update(id string, req *protocol.UpdateConnectionRequest) (protocol.ConnectionInfo, error) {
	existing, err := r.Get(id)
	if err != nil {
		return protocol.ConnectionInfo{}, err
	}
	now := time.Now().UTC()

	var passwordRef string
	if req.PasswordRef != nil {
		passwordRef = *req.PasswordRef
	} else {
		var ref sql.NullString
		if err := r.db.QueryRow("SELECT password_ref FROM connections WHERE id = ?", id).Scan(&ref); err != nil && err != sql.ErrNoRows {
			return protocol.ConnectionInfo{}, fmt.Errorf("read password_ref: %w", err)
		}
		passwordRef = ref.String
	}

	name := existing.Name
	host := existing.Host
	port := existing.Port
	database := existing.Database
	username := existing.Username
	options := existing.Options
	ssh := existing.SSHTunnel
	defSchema := existing.DefaultSchema
	readOnly := existing.ReadOnly
	group := existing.Group
	color := existing.Color

	if req.Name != nil {
		name = *req.Name
	}
	if req.Host != nil {
		host = *req.Host
	}
	if req.Port != nil {
		port = *req.Port
	}
	if req.Database != nil {
		database = *req.Database
	}
	if req.Username != nil {
		username = *req.Username
	}
	if req.Options != nil {
		options = req.Options
	}
	if req.SSHTunnel != nil {
		ssh = req.SSHTunnel
	}
	if req.DefaultSchema != nil {
		defSchema = *req.DefaultSchema
	}
	if req.ReadOnly != nil {
		readOnly = req.ReadOnly
	}
	if req.Group != nil {
		group = req.Group
	}
	if req.Color != nil {
		color = req.Color
	}

	optionsJSON, err := json.Marshal(options)
	if err != nil {
		return protocol.ConnectionInfo{}, fmt.Errorf("marshal options: %w", err)
	}
	var sshJSON []byte
	if ssh != nil {
		sshJSON, err = json.Marshal(ssh)
		if err != nil {
			return protocol.ConnectionInfo{}, fmt.Errorf("marshal ssh: %w", err)
		}
	}
	_, err = r.db.Exec(
		`UPDATE connections SET name=?, host=?, port=?, database=?, username=?, password_ref=?, options=?, ssh_tunnel=?, default_schema=?, read_only=?, "group"=?, color=?, updated_at=? WHERE id=?`,
		name, nullStr(host), nullInt(port), nullStr(database), nullStr(username),
		nullStr(passwordRef), string(optionsJSON), nullBytes(sshJSON), nullStr(defSchema), boolInt(readOnly),
		nullStr(derefStr(group)), nullStr(derefStr(color)),
		now.Format(time.RFC3339), id,
	)
	if err != nil {
		return protocol.ConnectionInfo{}, fmt.Errorf("update connection: %w", err)
	}
	return protocol.ConnectionInfo{
		ID:            id,
		Name:          name,
		Kind:          existing.Kind,
		Host:          host,
		Port:          port,
		Database:      database,
		Username:      username,
		Options:       options,
		SSHTunnel:     ssh,
		DefaultSchema: defSchema,
		ReadOnly:      readOnly,
		Group:         group,
		Color:         color,
		CreatedAt:     existing.CreatedAt,
		UpdatedAt:     now,
	}, nil
}

func (r *ConnectionRepository) Delete(id string) (bool, error) {
	res, err := r.db.Exec("DELETE FROM connections WHERE id = ?", id)
	if err != nil {
		return false, fmt.Errorf("delete connection: %w", err)
	}
	n, _ := res.RowsAffected()
	return n > 0, nil
}

// GetPasswordRef 返回连接存储的 password_ref；该引用不随 ConnectionInfo 下发（红线）。
func (r *ConnectionRepository) GetPasswordRef(id string) (string, error) {
	var ref sql.NullString
	err := r.db.QueryRow("SELECT password_ref FROM connections WHERE id = ?", id).Scan(&ref)
	if err == sql.ErrNoRows {
		return "", ErrNotFound
	}
	if err != nil {
		return "", fmt.Errorf("read password_ref: %w", err)
	}
	return ref.String, nil
}

type rowScanner interface {
	Scan(dest ...any) error
}

func scanConnection(s rowScanner) (protocol.ConnectionInfo, error) {
	var (
		id, name, kind                string
		host, database, username      sql.NullString
		port                          sql.NullInt64
		optionsJSON, created, updated string
		sshJSON                       sql.NullString
		defSchema                     sql.NullString
		readOnly                      int
		group, color                  sql.NullString
	)
	err := s.Scan(&id, &name, &kind, &host, &port, &database, &username,
		&optionsJSON, &sshJSON, &defSchema, &readOnly, &group, &color, &created, &updated)
	if err != nil {
		return protocol.ConnectionInfo{}, err
	}
	var options map[string]string
	if err := json.Unmarshal([]byte(optionsJSON), &options); err != nil {
		options = map[string]string{}
	}
	var ssh *protocol.SshTunnelConfig
	if sshJSON.Valid && sshJSON.String != "" {
		if err := json.Unmarshal([]byte(sshJSON.String), &ssh); err != nil {
			return protocol.ConnectionInfo{}, fmt.Errorf("parse ssh: %w", err)
		}
	}
	createdAt, _ := time.Parse(time.RFC3339, created)
	updatedAt, _ := time.Parse(time.RFC3339, updated)
	var ro *bool
	if readOnly != 0 {
		ro = &readOnlyFlag
	}
	return protocol.ConnectionInfo{
		ID:            id,
		Name:          name,
		Kind:          protocol.DatabaseKind(kind),
		Host:          host.String,
		Port:          int(port.Int64),
		Database:      database.String,
		Username:      username.String,
		Options:       options,
		SSHTunnel:     ssh,
		DefaultSchema: defSchema.String,
		ReadOnly:      ro,
		Group:         nullStrPtr(group),
		Color:         nullStrPtr(color),
		CreatedAt:     createdAt,
		UpdatedAt:     updatedAt,
	}, nil
}

func nullStr(s string) any {
	if s == "" {
		return nil
	}
	return s
}

func nullInt(v int) any {
	if v == 0 {
		return nil
	}
	return v
}

func nullBytes(b []byte) any {
	if len(b) == 0 {
		return nil
	}
	return string(b)
}

// boolInt 把可选布尔写成 0/1（nil 视为 false），供 read_only 列使用。
func boolInt(v *bool) int {
	if v != nil && *v {
		return 1
	}
	return 0
}

func derefStr(v *string) string {
	if v == nil {
		return ""
	}
	return *v
}

func nullStrPtr(v sql.NullString) *string {
	if !v.Valid {
		return nil
	}
	return &v.String
}

var readOnlyFlag = true
